test/test_dynamic_shapes_detector.py

```diff
@@ -0,0 +1,233 @@
+import torch
+import torch_xla
+import test_utils
+import unittest
+
+
+def set_max_allowed_traces(traces):
+  torch_xla._XLAC._dynamic_shape_detector_set_max_allowed_traces(traces)
+
+
+class TestDynamicShapeDetector(test_utils.XlaTestCase):
+
+  def _run_and_compare(self, f, optf=None, args=None):
+    """Run f and its torch_xla.compile wrapped version, comparing the equality
+    of their results.
+
+    If no optf is provided, we create a new one by wrapping it with
+    torch_xla.compile ourselves.
+    """
+    optf = optf or torch_xla.compile(f, detect_dynamic_shape=True)
+    args = args or []
+
+    out = f(*args)
+    optout = optf(*args)
+
+    self.assertEqual(out, optout)
+
+  def test_single(self):
+    # Test: trace a function once, when only one trace is allowed.
+
+    set_max_allowed_traces(1)
+
+    def foo(x):
+      return x + x
+
+    inp = torch.rand(10, device=torch_xla.device())
+    self._run_and_compare(foo, args=(inp,))
+
+  def test_many_traces(self):
+    # Test: multiple traces of a function.
+    #
+    # Steps 0~2 and 5: create new traces.
+    # Steps 3 and 4: ensure we have already traced these paths.
+
+    set_max_allowed_traces(4)
+
+    def foo(x, step):
+      r0 = x + x + x
+      r = r0 + x
+      if step in (0, 3):
+        return r + x
+      if step == (1, 4):
+        return r * 2
+      if step == 2:
+        return r * 4
+      return r0
+
+    optfoo = torch_xla.compile(foo, detect_dynamic_shape=True)
+    inp = torch.rand(10, device=torch_xla.device())
+
+    for i in range(6):
+      self._run_and_compare(foo, args=(inp, i))
+
+  def test_trace_limit_exceeded_different_input_shape(self):
+    # Test: catch trace limit exceeded error when running the function with a
+    # function with different shape.
+
+    set_max_allowed_traces(1)
+
+    def foo(x):
+      return x + x
+
+    optfoo = torch_xla.compile(foo, detect_dynamic_shape=True)
+
+    inp1 = torch.rand(10, device=torch_xla.device())
+    self._run_and_compare(foo, optfoo, args=(inp1,))
+
+    msg = """\
+torch_xla/csrc/dynamic_shape_detector.cpp:47 : Maximum number of different traces allowed per function exceeded: 1
```

Comment:
`cpp:47` is too specified, we can remove the line number and file name from the check to make it more general.